x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -391,21 +399,52 @@ private void startComputeOnRemoteClusters(
                     queryPragmas.exchangeBufferSize(),
                     esqlExecutor,
                     refs.acquire().delegateFailureAndWrap((l, unused) -> {
-                        var remoteSink = exchangeService.newRemoteSink(rootTask, childSessionId, transportService, cluster.connection);
-                        exchangeSource.addRemoteSink(remoteSink, true, queryPragmas.concurrentExchangeClients(), ActionListener.noop());
                         var remotePlan = new RemoteClusterPlan(plan, cluster.concreteIndices, cluster.originalIndices);
                         var clusterRequest = new ClusterComputeRequest(cluster.clusterAlias, childSessionId, configuration, remotePlan);
-                        var clusterListener = ActionListener.runBefore(
-                            computeListener.acquireCompute(cluster.clusterAlias()),
-                            () -> l.onResponse(null)
+
+                        // Create group task for this cluster
+                        CancellableTask groupTask = createGroupTask(rootTask, clusterRequest.getDescription());
+                        CountDown countDown = new CountDown(2);
+                        // The group is done when both the sink and the cluster request are done
+                        Runnable finishGroup = () -> {
+                            if (countDown.countDown()) {
```

Comment:
Are we always guaranteed that both the sink and cluster requests will finish and call this finishGroup Runnable? Can one ever finish but not the other?

What about if we are in a long-running query and clusterX is skip_un=true and has an earlier failure. Will this cleanup code run soon thereafter or does it have to wait until the full long query has finished?

Reply:
`runAfter`s below ensure that both branches will call the finisher.